src/components/RegisterPopupup/RegisterPopupup.js

```diff
@@ -0,0 +1,126 @@
+import PopupWithForm from "../PopupWithForm/PopupWithForm.js";
+import React from "react";
+import {useSelector} from "react-redux";
+import {register} from "../../utils/auth.js";
+import {txtErr} from "../../utils/constants.js";
+
+const RegisterPopupup = ({
+  onChangeInput,
+  handlePopupMouseDown,
+  handlePopupToggleView,
+  isInputHaveKey,
+  onFormSubmitted
+}) => {
+  const {
+    type,
+    isOpen,
+    title,
+    isFormVaild,
+    btnSetting,
+    bottomLink,
+    serverError,
+    inputs
+  } = useSelector((state) => state.fromSettingModule);
+  const settingPopupWithForm = {
+    type,
+    isOpen,
+    title,
+    isFormVaild,
+    btnSetting,
+    inputs,
+    bottomLink
+  };
+  const currType = "register";
+
+  const handleNewUserSubmit = (e) => {
+    e.preventDefault();
+    onFormSubmitted({isDone: false});
+    register({
+      email: inputs.emailAddress.inputVal.toLowerCase().toString(),
+      password: inputs.userPassword.inputVal,
+      name: inputs.userName.inputVal
+    })
+      .then(() => {
+        onFormSubmitted({
+          isDone: true,
+          btnSetting: {txt: "Loading...", isDisable: true}
+        });
+        handlePopupToggleView("signup_success");
+      })
+      .catch((err) => {
+        onFormSubmitted({
+          isDone: true,
+          btnSetting: {txt: "Sign up", isDisable: false},
+          err: err.message || txtErr
+        });
+      });
+  };
+  return (
+    <PopupWithForm
+      handlePopupMouseDown={handlePopupMouseDown}
+      handleSubmit={handleNewUserSubmit}
+      settingPopupWithForm={settingPopupWithForm}
+      currType={currType}
+      handlePopupToggleView={handlePopupToggleView}
+    >
+      <label className="popup__label">Email</label>
+      <input
+        className="popup__input"
+        autoFocus
+        onChange={(e) => {
+          onChangeInput(e);
+        }}
+        type="email"
+        placeholder="Enter email"
+        name="emailAddress"
+        title="Invalid email address"
+        pattern="[^@ \t\r\n]+@[^@ \t\r\n]+\.[^@ \t\r\n]+"
+        value={isInputHaveKey({key: "emailAddress", subKey: "inputVal"})}
+        required
+      />
+      <span className="popup__input-error">
+        {isInputHaveKey({key: "emailAddress", subKey: "inputMsgVaild"})}
+      </span>
+      <label className="popup__label">Password</label>
+      <input
+        className="popup__input"
+        type="password"
+        value={isInputHaveKey({key: "userPassword", subKey: "inputVal"})}
+        name="userPassword"
+        minLength="6"
+        onChange={(e) => {
+          onChangeInput(e);
+        }}
+        placeholder="Enter password"
+        required
+      />
+      <span className="popup__input-error">
+        {isInputHaveKey({key: "userPassword", subKey: "inputMsgVaild"})}
```

Comment:
**Might be better**

If it’s interesting for you here is a custom `validation hook`.

You should put it in a separate file  `useFormAndValidation.js` in folder `hooks` and then import it to the component that needs validation:

```js

import {useState, useCallback} from 'react';

export function useFormAndValidation() {
  const [ values, setValues ] = useState({});
  const [ errors, setErrors ] = useState({});
  const [ isValid, setIsValid ] = useState(true);

  const handleChange = (e) => {
    const {name, value} = e.target
    setValues({...values, [name]: value });
    setErrors({...errors, [name]: e.target.validationMessage});
    setIsValid(e.target.closest('form').checkValidity());
  };



  const resetForm = useCallback((newValues = {}, newErrors = {}, newIsValid = false) => {
    setValues(newValues);
    setErrors(newErrors);
    setIsValid(newIsValid);
  }, [setValues, setErrors, setIsValid]);

  return { values, handleChange, errors, isValid, resetForm, setValues, setIsValid };
}


```

And only one line of code enables validation:

```js

  const {values, handleChange, errors, isValid, setValues, resetForm} = useFormAndValidation()

```

So, you **don’t need** to make functions for handling inputs, storing values, resetting the form. Everything is already done.